src/psyclone/psyir/nodes/scoping_node.py

```diff
@@ -119,11 +125,57 @@ def _refine_copy(self, other):
         :type other: :py:class:`psyclone.psyir.node.Node`
 
         '''
-        super(ScopingNode, self)._refine_copy(other)
+        # Reorganise the symbol table construction to occur before we add
+        # the children.
         self._symbol_table = other.symbol_table.deep_copy()
-        # pylint: disable-next=protected-access
         self._symbol_table._node = self  # Associate to self
 
+        # Remove symbols corresponding to Routines that are contained in this
+        # ScopingNode. These symbols will be added automatically by the Routine
+        # objects when we add them to our child list in the super refine_copy
+        # call.
+        # We have to import Routine here to avoid a circular dependency.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.routine import Routine
+        for node in other.walk(Routine):
+            try:
+                if isinstance(self._symbol_table.lookup(node.name),
+                              RoutineSymbol):
+                    self._symbol_table.remove(
+                            self._symbol_table.lookup(node.name))
+            except KeyError:
+                pass
+            except ValueError:
+                # If remove fails for the RoutineSymbol then we have to force
+                # the removal of the symbol.
+                symbol = self._symbol_table.lookup(node.name)
+                norm_name = self._symbol_table._normalize(symbol.name)
+                self._symbol_table._symbols.pop(norm_name)
```

Comment:
If we have to force it anyway, it could be better to replace:
`self._symbol_table.remove(self._symbol_table.lookup(node.name))`
with:
```
# Replace it manually to avoid checking if there are still references to this symbol,
# those references will be also replaced below.
symbol = self._symbol_table.lookup(node.name)
norm_name = self._symbol_table._normalize(symbol.name)
self._symbol_table._symbols.pop(norm_name)
```
And remove the "except ValueError", since it's the expected path anyway.

Reply:
Yeah - I was initially concerned whether we had to fix any `Call` nodes to point to this symbol, but that has never happened (I had an `assert False` to check for this) so I think that this is safe. Done.